packages/lit-html/src/directives/spread.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {noChange, nothing, _Σ as litHtmlPrivate} from '../lit-html.js';
+import {
+  directive,
+  Directive,
+  DirectiveParameters,
+  PartInfo,
+  PartType,
+  Part,
+  ElementPart,
+} from '../directive.js';
+
+const {
+  _AttributePart: AttributePartImpl,
+  _PropertyPart: PropertyPartImpl,
+  _BooleanAttributePart: BooleanAttributePartImpl,
+  _EventPart: EventPartImpl,
+  // _ElementPart: ElementPartImpl,
+} = litHtmlPrivate;
+
+export interface SpreadValues {
+  readonly [name: string]: unknown;
+}
+
+class SpreadDirective extends Directive {
+  _previousValues: Set<string> = new Set();
+  _parts = new Map<string, Part>();
+
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    if (partInfo.type !== PartType.ELEMENT) {
+      throw new Error(
+        'The `spread` directive must be used al element expression'
+      );
+    }
+  }
+
+  render(_values: SpreadValues) {
```

Comment:
I can imagine use cases where it'd be nice to let `values` be nullable, to avoid users from needing to always make a throwaway object in the alternate case:
```html
<div ${spread(maybeGotStuff ?? {})}></div>
```
or
```html
<div ${spread(shouldRenderStuff ? stuff : {})}></div>
```

Again, we don't do this in `classMap`/`styleMap`, but I think it's more likely to have patterns where entire bags are forwarded in (e.g. "`formControlAttributes`") vs. being generated by the user (otherwise the user could have just written the attribute bindings themselves).

But it does sort of beg the question what we expect `spread` to actually be used for in the OSS world.

Reply:
Agreed, as per the other thread.

I imagine this form of spread to be a lower level things that's often wrapped for better DX. An element might receive a bag of property or event bindings and need to add the correct prefix. `attr` will wrap this. Custom attribute-merging strategies will... do something. We need to make sure that's possible.